packages/juce-blueprint/esm/Blueprint.js

```diff
@@ -22,52 +22,85 @@ export function Image(props) {
   return React.createElement('Image', props, props.children);
 }
 
-export function Canvas(props) {
-  if (props.hasOwnProperty('onDraw')) {
-    let userOnDraw = props.onDraw;
-    let wrappedDraw = function(ctx) {
-      Object.defineProperty(ctx, 'fillStyle', {
-        enumerable: false,
-        configurable: false,
-        get: function() {
-          return 'Not Supported';
-        },
-        set: function(value) {
-          this.__setFillStyle(value);
-        }
-      });
-
-      Object.defineProperty(ctx, 'strokeStyle', {
-        enumerable: false,
-        configurable: false,
-        get: function() {
-          return 'Not Supported';
-        },
-        set: function(value) {
-          this.__setStrokeStyle(value);
-        }
-      });
-
-      Object.defineProperty(ctx, 'lineWidth', {
-        enumerable: false,
-        configurable: false,
-        get: function() {
-          return 'Not Supported';
-        },
-        set: function(value) {
-          this.__setLineWidth(value);
-        }
-      });
-
-      return userOnDraw(ctx);
+export function bindCanvasContextProperties(ctx) {
+  Object.defineProperty(ctx, 'fillStyle', {
+    enumerable: false,
+    configurable: false,
+    get: function() {
+      return 'Not Supported';
+    },
+    set: function(value) {
+      this.__setFillStyle(value);
+    }
+  });
+
+  Object.defineProperty(ctx, 'strokeStyle', {
+    enumerable: false,
+    configurable: false,
+    get: function() {
+      return 'Not Supported';
+    },
+    set: function(value) {
+      this.__setStrokeStyle(value);
+    }
+  });
+
+  Object.defineProperty(ctx, 'lineWidth', {
+    enumerable: false,
+    configurable: false,
+    get: function() {
+      return 'Not Supported';
+    },
+    set: function(value) {
+      this.__setLineWidth(value);
+    }
+  });
+}
+
+export class Canvas extends Component {
+  constructor(props) {
+    super(props);
+    this._onMeasure = this._onMeasure.bind(this);
+
+    this.state = {
+      width: 0,
+      height: 0
     };
+  }
 
-    return React.createElement('CanvasView', Object.assign({}, props, {
-      onDraw: wrappedDraw,
-    }), props.children);
+  _onMeasure(width, height) {
+    this.setState({
+      width: width,
+      height: height
+    });
   }
 
-  return React.createElement('CanvasView', props, props.children);
+  render() {
+    let props = this.props;
+    let userOnDraw = props.onDraw;
+
+    const {width, height} = this.state
+
+    let canvas = React.createElement('CanvasView', props, props.children);
+
+    if (userOnDraw) {
+      let wrappedDraw = function (ctx) {
+        bindCanvasContextProperties(ctx);
+        if (props.autoclear) ctx.clearRect(0, 0, width, height);
+        return userOnDraw(ctx);
+      }
+
+      canvas = React.createElement('CanvasView', Object.assign({}, props, {
+        onDraw: wrappedDraw
+      }), props.children);
+    }
+
+    return (
+        <View {...this.props} onMeasure={this._onMeasure}>
```

Comment:
Hmm... you shoudn't need to wrap this with a View. As long as you put the `onMeasure` prop on the canvas you should get the callback.

Reply:
Ah, got it! Check out the `CanvasView::resized()` method– that override should call the base class `View::resized()` and then do its `ctx->setSize` (as we do in the CanvasView::paint). The View base class does all the usual business of calling the `onMeasure` callback and painting things like borders and background colors

Reply:
Ahhhhhaaaa. Ha of course. I was confused as to why the CanvasView's onMeasure prop wasn't working. Let me try fixing that up now.

Reply:
Hmm. This doesn't appear to be working yet when trying to bind the outer component onMeasure. I'm missing how to make this work. 

```
     let onMeasure = this._onMeasure;

      canvas = React.createElement('CanvasView', Object.assign({}, props, {
        onDraw: wrappedDraw,
        onMeasure: onMeasure
      }), props.children);
```

Reply:
Hmm, I think you'll basically want this?

```js
class CanvasView extends React.Component {
  constructor(props) {
    super(props);

    this._onMeasure = this._onMeasure.bind(this);
    this._onDraw = this._onDraw.bind(this);

    this.state = {
      width: 0,
      height: 0,
    };
  }

  _onMeasure(w, h) {
    this.setState({ width: w, height: h });
    if (typeof this.props.onMeasure === 'function')
      this.props.onMeasure(w, h);
  }

  _onDraw(ctx) {
    if (typeof this.props.onDraw === 'function') {
      bindCtx(ctx);
      if (this.props.autoclear)
        ctx.clearRect(0, 0, this.state.width, this.state.height);
      this.props.onDraw(ctx);
    }
  }
  render() {
    return React.createElement('CanvasView', Object.assign({}, this.props, {
      onDraw: this._onDraw,
      onMeasure: this._onMeasure,
    }, this.props.children);
  }
}
```

Reply:
@nick-thompson, maybe ignore me ... think I've sorted

Reply:
@nick-thompson, ha yeah OK I fixed but yours is nicer than mine. I need a little js refresh session at some point I think ...

Reply:
Curses no that isn't working. Not getting any repaints now. Will investigate.

Reply:
😕 confusing. Yea, I'd have to dig into it... that _should_ work.. I think

Reply:
Yeah no worries. I'll sort it, something daft I expect

Reply:
Works with this but I'll try to make it fit your structure more as it's nicer:

```
export class Canvas extends Component {
  constructor(props) {
    super(props);
    this._onMeasure = this._onMeasure.bind(this);

    this.state = {
      width: 0,
      height: 0
    };
  }

  _onMeasure(width, height) {
    console.log('canvas on measure');
    this.setState({
      width: width,
      height: height
    });
  }

  render() {
    let props = this.props;
    let userOnDraw = props.onDraw;

    const {width, height} = this.state

    let canvas = React.createElement('CanvasView', props, props.children);

    if (userOnDraw) {
      let wrappedDraw = (ctx) => {
        bindCanvasContextProperties(ctx);
        if (props.autoclear) ctx.clearRect(0, 0, width, height);
        return userOnDraw(ctx);
      }

      canvas = React.createElement('CanvasView', Object.assign({}, props, {
        onDraw: wrappedDraw,
        onMeasure: this._onMeasure
      }), props.children);
    }

    return canvas;
  }
}
```

Reply:
That's super odd... they feel basically identical. I'm cool with how you have it there though honestly! The only thing I would add is in the _onMeasure hook, check if the user tried to supply their own `onMeasure` and if so, call it there

Reply:
Right ... I'm determined to workout why it isn't working

Reply:
Pfft. Stumped. Possible it's anything to do with my usage of the Canvas component?

```
          <View
              {...this.props}
              onMeasure={this._onMeasure}
              onMouseDown={this._onMouseDown}
              onMouseDrag={this._onMouseDrag}>
            <Canvas autoclear={true} {...styles.canvas} onDraw={this._drawSlider} />
            {this.props.children}
          </View>
```

Reply:
I doubt it. You're not getting any errors right? The paint's just not happening?

Reply:
Nope nothing. This works fine:

```
export class Canvas extends Component {
  constructor(props) {
    super(props);
    this._onMeasure = this._onMeasure.bind(this);

    this.state = {
      width: 0,
      height: 0
    };
  }

  _onMeasure(width, height) {
    this.setState({
      width: width,
      height: height
    });
  }

  render() {
    let props = this.props;
    const {width, height} = this.state

    let wrappedDraw = (ctx) => {
      bindCanvasContextProperties(ctx);
      if (props.autoclear) ctx.clearRect(0, 0, width, height);
      if (typeof props.onDraw === 'function') props.onDraw(ctx);
    }

    let wrappedOnMeasure = (width, height) => {
      this._onMeasure(width, height);
      if (typeof props.onMeasure === 'function') props.onMeasure(width, height);
    }

    return React.createElement('CanvasView', Object.assign({}, props, {
      onDraw: wrappedDraw,
      onMeasure: wrappedOnMeasure
    }), props.children);
  }
}
```

How strange.

Reply:
@nick-thompson, I'm going to call it quits for the day so I've pushed up the only working version I can get in https://github.com/nick-thompson/blueprint/pull/56/commits/1e351ae97ae5b4785f035d5d836982c1b4b32c60 incase you want to take a look and see if you can workout why we can't get your implementation to work. I don't understand it myself.

Reply:
Took a pass this morning, I've got it working on my end! Only needed a few small changes: https://gist.github.com/nick-thompson/d28a7fcd334fe0c2eb47ff4636649da6 (the unique_ptr stuff is just because I'm on an older version of juce, ignore those)

You should be able to apply that diff to your branch locally and give it a run. Let me know if that works for you

Reply:
@nick-thompson, heh. Good catch with std::max!

Strangely still the same issue for me so must be something my end. Were you testing this in an example project?

Reply:
Nice catch with std::max. 

So I must be missing something but I can only make your gist work if I wrap this._onDraw and this._onMeasure. Some JS issue I'm ignorant of here. I might test on mac with my project just in case this is some total weird platform specific thing. 

This works for me:

```
export class Canvas extends Component {
  constructor(props) {
    super(props);

    this._onMeasure = this._onMeasure.bind(this);
    this._onDraw    = this._onDraw.bind(this);

    this.state = {
      width: 0,
      height: 0
    };
  }

  _onMeasure(width, height) {
    this.setState({
      width: width,
      height: height
    });

    if (typeof this.props.onMeasure === 'function')
      this.props.onMeasure(width, height);
  }

  _onDraw(ctx) {
    if (typeof this.props.onDraw === 'function') {
      bindCanvasContextProperties(ctx);

      if (this.props.autoclear) {
        ctx.clearRect(0, 0, this.state.width, this.state.height);
      }

      this.props.onDraw(ctx);
    }
  }

  render() {
    let wrappedDraw = (ctx) => {
      this._onDraw(ctx);
    }

    let wrappedOnMeasure = (width, height) => {
      this._onMeasure(width, height);
    }

    return React.createElement('CanvasView', Object.assign({}, this.props, {
      onDraw: wrappedDraw,
      onMeasure: wrappedOnMeasure
    }), this.props.children);
  }
}
```

Reply:
Does seem to be an issue binding this in Object.assign() as the below also works. Not clear on why I'm the only one seeing this though. 

```
  render() {
    return React.createElement('CanvasView', Object.assign({}, this.props, {
      onDraw: (ctx) => { this._onDraw(ctx) },
      onMeasure: (width, height) => { this._onMeasure(width, height) }
    }), this.props.children);
  }

```

Reply:
Wow, no kidding. Ok, thanks for the update. This sounds like a bug in duktape to be honest with you. Wondering if we'll fix the issue just by updating our duktape installation...

Let's use the code from your comment:

```
  render() {
    return React.createElement('CanvasView', Object.assign({}, this.props, {
      onDraw: (ctx) => { this._onDraw(ctx) },
      onMeasure: (width, height) => { this._onMeasure(width, height) }
    }), this.props.children);
  }
```
And can you also add a todo comment?
```
// TODO: Check this for bug in duktape
```

That at least will unblock this PR from going in, and we can revisit later if we find anything in duktape. Sound good?

Reply:
Yeah that's all I can boil it down to, I'm testing this on Ubuntu so maybe a duktape on linux quirk. Will add the TODO. Thanks Nick!

Reply:
@nick-thompson, have rebased and pushed up changes with you advised comment in https://github.com/nick-thompson/blueprint/commit/9540539efc1763be4bf18024768abeae0b09606c. Once this is in I will fix up https://github.com/nick-thompson/blueprint/pull/58 and then I think we're at a good point to start adding some Canvas/drawing docs and example apps etc.

 I think I'll raise an issue for the Canvas docs/examples which we can use to track to work and discuss ideas. Sound good?

Reply:
Awesome, sounds great! I'll take one last look here now and merge if all's good